pennylane/qnodes/passthru.py

```diff
@@ -0,0 +1,145 @@
+    (2) indexing/slicing, and possibly (3) iteration, as these are the things qfuncs expect of
+    array-like parameters.
+    """
+    def __init__(self, func, device, properties=None):
+        super().__init__(func, device, mutable=True, properties=properties)
+
+    def __repr__(self):
+        """String representation."""
+        detail = "<PassthruQNode: device='{}', func={}, wires={}>"
+        return detail.format(self.device.short_name, self.func.__name__, self.num_wires)
+
+    def _set_variables(self, args, kwargs):
+        # do nothing, since we do not use VariableRefs
+        pass
+
+    def _construct(self, args, kwargs):
+        """Construct the quantum circuit graph by calling the quantum function.
+
+        Like :class:`.BaseQNode._construct`, but does not use VariableRefs.
+
+        PassthruQNode does not have a Jacobian method, so it does not HAVE to use VariableRefs or scalar linear indexing of input parameters.
+        Two options:
+        1. Use VariableRefs anyway, re-use most BaseQNode methods.
+           Problem: after evaluating the VariableRefs, stacking sliced/indexed Tensors in Operation.parameters should somehow result in a Tensor, not an object array.
+        2. Do not use VariableRefs, call the qfunc each time :meth:`PassthruQNode.evaluate` is called (always mutable).
+           Problem: tensornet_tf requires variable_deps?
+
+
+        TODO rethink output_conversion? should require device to return things in a fixed form, but either as arrays or as AD Tensors, do conversion in interface (if necessary...)
+        """
+        # temporary queues for operations and observables
+        self.queue = []  #: list[Operation]: applied operations
+        self.obs_queue = []  #: list[Observable]: applied observables
+
+        # set up the context for Operator entry
+        if qml._current_context is None:
+            qml._current_context = self
+        else:
+            raise QuantumFunctionError(
+                "qml._current_context must not be modified outside this method."
+            )
+        try:
+            # turn off domain checking since PassthruQNode qfuncs can take any class as input
+            pennylane.operation.Operator.do_check_domain = False
+            # generate the program queue by executing the quantum circuit function
+            res = self.func(*args, **kwargs)
+        finally:
+            qml._current_context = None
+            pennylane.operation.Operator.do_check_domain = True
+
+        # check the validity of the circuit
+        self._check_circuit(res)
+        del self.queue
+        del self.obs_queue
+
+        # no VariableRefs, self.variable_deps is empty!
+        # generate the DAG
+        self.circuit = pennylane.circuit_graph.CircuitGraph(self.ops, self.variable_deps)
+
+        # check for operations that cannot affect the output
+        if self.properties.get("vis_check", False):
```

Comment:
I realize this is just code ported over from the base class, but what is the goal of this check? Is it just for debugging?

Reply:
btw since the code is duplicated, might make sense to make it it's own method in the parent class, and just call `self._vis_check()`?

Reply:
In the long run, it would be great if we could just get a call to `super.__construct__` working here. Would require pulling out some code from the "Base" class that is no longer a shared base implementation. 

In other words, rather than writting a stripped down `_construct` in `PassThruQNode`, shouldn't what is happening in `BaseQNode` be enough for `PassThruQnode` to work, and any "Black box" QNodes should have additional code to handle their special needs?

Reply:
I moved the visibility check to `BaseQNode`. Its purpose is to warn users if their qfunc makes no sense (having gates that cannot influence the output).

When it comes to construction PassthruQNode works quite differently from the other QNodes, it does not use VariableRefs for example, so I don't think it'd be very useful to force it to use the base class _construct method. But I can look into it while optimizing the QNodes.

Reply:
thumb

Reply:
Not really part of this PR, but this feature still seems weird to me. I'm assuming this is a developer/debugging feature, but I don't understand the implementation.

Why would we want a function which raises an exception if there are gates which do not affect the output? 
I can think of some use-cases where I might have a circuit whose output is not influenced by a particular gate, e.g., building up a circuit step-by-step by hand and checking the outputs at each step (we had to do this recently while debugging a black-box plugin). I would still want that to execute without error (perhaps telling me which gates are disconnected). 

Wouldn't it be better to have a function which returns a set of "invisible" gates, rather than one that just errors?

Reply:
By default the visibility check is off, but

> Wouldn't it be better to have a function which returns a set of "invisible" gates, rather than one that just errors?

it probably would, and maybe throw a warning.

Reply:
I added the function, but kept the error for now, since the check is off by default. The behavior can be changed later.